src/shared/varlink.c

```diff
@@ -0,0 +1,1962 @@
+                v->time_event_source = sd_event_source_unref(v->time_event_source);
+        }
+
+        if (v->quit_event_source) {
+                (void) sd_event_source_set_enabled(v->quit_event_source, SD_EVENT_OFF);
+                v->quit_event_source = sd_event_source_unref(v->quit_event_source);
+        }
+
+        if (v->defer_event_source) {
+                (void) sd_event_source_set_enabled(v->defer_event_source, SD_EVENT_OFF);
+                v->defer_event_source = sd_event_source_unref(v->defer_event_source);
+        }
+}
+
+static Varlink* varlink_destroy(Varlink *v) {
+        if (!v)
+                return NULL;
+
+        varlink_detach_event_sources(v);
+
+        varlink_server_unref(v->server);
+
+        safe_close(v->fd);
+
+        free(v->input_buffer);
+        free(v->output_buffer);
+
+        json_variant_unref(v->current);
+        json_variant_unref(v->reply);
+
+        sd_event_unref(v->event);
+
+        return mfree(v);
+}
+
+DEFINE_TRIVIAL_REF_UNREF_FUNC(Varlink, varlink, varlink_destroy);
+
+static int varlink_write(Varlink *v) {
+        ssize_t n;
+
+        assert(v);
+
+        if (!VARLINK_STATE_IS_CONNECTED(v->state))
+                return 0;
+        if (v->output_buffer_size == 0)
+                return 0;
+        if (v->write_disconnected)
+                return 0;
+
+        assert(v->fd >= 0);
+
+        n = write(v->fd, v->output_buffer + v->output_buffer_index, v->output_buffer_size);
+        if (n < 0) {
+                if (errno == EAGAIN)
+                        return 0;
+
+                if (ERRNO_IS_DISCONNECT(errno)) {
+                        /* If we get informed about a disconnect on write, then let's remember that, but not
+                         * act on it just yet. Let's wait for read() to report the issue first. */
+                        v->write_disconnected = true;
+                        return 0;
+                }
+
+                return -errno;
+        }
+
+        v->output_buffer_size -= n;
+
+        if (v->output_buffer_size == 0)
+                v->output_buffer_index = 0;
+        else
+                v->output_buffer_index += n;
+
+        v->timestamp = now(CLOCK_MONOTONIC);
+        return 1;
+}
+
+static int varlink_read(Varlink *v) {
+        size_t rs;
+        ssize_t n;
+
+        assert(v);
+
+        if (!IN_SET(v->state, VARLINK_AWAITING_REPLY, VARLINK_CALLING, VARLINK_IDLE_SERVER))
+                return 0;
+        if (v->current)
+                return 0;
+        if (v->input_buffer_unscanned > 0)
+                return 0;
+
+        if (v->input_buffer_size >= VARLINK_BUFFER_MAX)
+                return -ENOBUFS;
+
+        assert(v->fd >= 0);
+
+        if (v->input_buffer_allocated <= v->input_buffer_index + v->input_buffer_size) {
+                size_t add;
+
+                add = MIN(VARLINK_BUFFER_MAX - v->input_buffer_size, VARLINK_READ_SIZE);
```

Comment:
input_buffer_index is set to 0 if it is not 0 further down: the newly allocated buffer will always start at byte zero and hence we only care about the actual size of the data in the buffer for the size estimations for the new allocation, not whether that data is located in the middle of the buffer or not...

i.e. the actually relevant data in the buffer starts at `input_buffer_index` and ends at `input_buffer_index + input_buffer_size`, and after resizing the buffer the relevant data always starts at `0` and ends at `0 + input_buffer_size` and `input_buffer_index` is set to 0 hence.

Reply:
> I must say I'm not fond of how this buffer stuff is being done, it's annoying to read and seems likely to contain bugs. It feels like there should be a better, more elegant way. Furthermore this seems like something that gets reimplemented over and over again, I've reviewed stuff like this so many times always slightly different and there should just be some generic shared thoroughly debugged piece dealing with this.

You are right about that, such code is seen frequently. In `casync` there's a concept of a ReallocBuffer which is supposed to abstract this kinda of growing fifo buffer. I figure we could sooner or later adopt a similar scheme in systemd, but right now I think it would not be easy to do port sd-bus at least to such a generic implementation since it deals with fds, creds, memfds, in the data stream. 

So I think adding a generalization for this makes sense, but maybe do that after this PR?